### PR TITLE
Compute static fingerprint in the pipgraph fragments

### PR DESCRIPTION
**Pull request by erickulcyk**
Currently, computing pip graph fingerprint is done as the graph is being built in pipgraph.builder.  This is expensive, and so we move it to be done at the time of pip graph fragment creation.  A new field is added in pip, and if it exists, the value of it is used instead of recomputing.  The one caveat is if the pip graph fragments uses sealled directories from another fragment.  This is left as a todo because Office currently does not exercise this functionality.

### CODE REVIEW

File: Public/Src/Engine/Scheduler/Graph/GraphFragmentBuilderTopSort.cs [hunk 1]
```diff
@@ -29,12 +32,48 @@ public class GraphFragmentBuilderTopSort : GraphFragmentBuilder
...
         /// <summary>
         /// Creates an instance of <see cref="GraphFragmentBuilder"/>.
         /// </summary>
         public GraphFragmentBuilderTopSort(LoggingContext loggingContext, PipExecutionContext pipExecutionContext, IConfiguration configuration) : base(loggingContext, pipExecutionContext, configuration)
         {
+            var extraFingerprintSalts = new ExtraFingerprintSalts(
+                configuration,
+                PipFingerprintingVersion.TwoPhaseV2,
+                string.Empty,
+                searchPathToolsHash: null);
```
Comment (erickulcyk) [inline Public/Src/Engine/Scheduler/Graph/GraphFragmentBuilderTopSort.cs]: todo: figure this out.

File: Public/Src/Pips/Dll/Operations/Pip.cs [hunk 2]
```diff
@@ -195,6 +206,16 @@ internal void Serialize(PipWriter writer)
             writer.Write((byte)PipType);
             writer.Start(GetType());
             InternalSerialize(writer);
+            if (StaticFingerprint.Length >0)
```
Comment (ljcollins25) [inline Public/Src/Pips/Dll/Operations/Pip.cs]: >>0 [](start = 41, length = 2)

Spacing.

**Comment by ljcollins25** [comment]
>     public class GraphFragmentBuilderTopSort : GraphFragmentBuilder

Why isn't this logic in the base type? Also, it should be configurable.

---
Refers to: Public/Src/Engine/Scheduler/Graph/GraphFragmentBuilderTopSort.cs:30 in a73706d. [](commit_id = a73706d32781e3c05ad81d63f74c98e9dc87cf15, deletion_comment = False)

**Comment by erickulcyk** [comment]
>     public class GraphFragmentBuilderTopSort : GraphFragmentBuilder

It needs the graph that is built up in this class, so it can't be in the base class.  Perhaps this class should be renamed so it implies more than top sort.

---
In reply to: [527708247](https://github.com/microsoft/BuildXL/pull/827#issuecomment-527708247) [](ancestors = 527708247)

---
Refers to: Public/Src/Engine/Scheduler/Graph/GraphFragmentBuilderTopSort.cs:30 in a73706d. [](commit_id = a73706d32781e3c05ad81d63f74c98e9dc87cf15, deletion_comment = False)

File: Public/Src/Engine/Scheduler/Graph/GraphFragmentBuilderTopSort.cs [hunk 3]
```diff
@@ -29,12 +32,48 @@ public class GraphFragmentBuilderTopSort : GraphFragmentBuilder
...
+                PipFingerprintingVersion.TwoPhaseV2,
+                configuration.Cache.CacheSalt,
+                new DirectoryMembershipFingerprinterRuleSet(configuration, pipExecutionContext.StringTable).ComputeSearchPathToolsHash());
+
+            m_pipStaticFingerprinter = new PipStaticFingerprinter(
+                pipExecutionContext.PathTable,
+                GetSealDirectoryFingerprint,
+                GetDirectoryProducerFingerprint,
+                extraFingerprintSalts,
+                PathExpander.Default)
```
Comment (ljcollins25) [inline Public/Src/Engine/Scheduler/Graph/GraphFragmentBuilderTopSort.cs]: >PathExpander.Default [](start = 16, length = 20)

I think you need a real mount path expander here. System mounts get tokenized with the real mount path expander.

**Pull request merged**